yjit/src/codegen.rs

```diff
@@ -1417,7 +1417,7 @@ fn guard_object_is_heap(
     asm.cmp(object, Qfalse.into());
     asm.je(Target::side_exit(counter));
 
-    if object_type.diff(Type::UnknownHeap) != TypeDiff::Incompatible {
+    if !object_type.is_heap() {
         asm.ctx.upgrade_opnd_type(object_opnd, Type::UnknownHeap);
     }
```

Comment:
```suggestion
    asm.ctx.upgrade_opnd_type(object_opnd, Type::UnknownHeap);
```

We have an early return for `is_heap()`, so this `if` statement always holds.